### DISCUSSION TITLE
Do I need connection pooling for MongoDB and Mongoose?

### DISCUSSION THREAD

**Discussion by Zeng1998** [vercel/vercel#6596] [Help] [answered]
> Rather than opening a connection with every request, connection pooling allows us to designate a single "pooler" that keeps an active connection to the database. When a user invokes a Serverless Function that would read from the database, it will first search for an already available connection instead of creating a new connection. For example, you might use PgBouncer for connection pooling.

What about this in the document is brief.

I use node.js and mongooes, I just want to deploy a CURD example with mongodb. 

But I cannot find any detailed example, or is that impossible?

I find out something on Internet which like that's possible in AWS lambda.

**Comment by leerob** [comment, marked as answer]
Hey! If you want to use MongoDB with Mongoose, we have an example created with it set up to handle connections in a serverless environment - definitely possible!

https://github.com/vercel/next.js/tree/canary/examples/with-mongodb-mongoose

**Reply by Zeng1998** [reply]
> Hey! If you want to use MongoDB with Mongoose, we have an example created with it set up to handle connections in a serverless environment - definitely possible!
> 
> https://github.com/vercel/next.js/tree/canary/examples/with-mongodb-mongoose

Hey, thanks first, I found this and try after I posting this.

I copied the [dbConnect.js](https://github.com/vercel/next.js/blob/canary/examples/with-mongodb-mongoose/lib/dbConnect.js) and used it in a serverless function, like this

```js
import dbConnect from "./dbConnect"
const {Game} = require("./models");

app.get("/api/test", async (req, res) => {
    await dbConnect()
    let r=await Game.find({});
    res.json({
        res:r,
    });
})
```

I add  `console.log("cache")`  in the code to see if it use cache.
Then I make a concurrent requests test and it seems to work. But when I see the monitor of the [mongodb cloud](https://cloud.mongodb.com/), I found that there are too many( more than 100 ) connections at that time.

So I want to know what problem is it?

In brief:

- Is this normal? (too many connections in MongoDB monitor but cache seems to success in Vercel)
- If not, is it the issue with MongoDB cloud? And are there any others who use serverless functions with MongoDB cloud can help me, please.

**Reply by willshulman** [reply]
Hi Zeng. Do you have a sense of what are the maximum number concurrent requests you have open at a time? Do you see the 100 number decay some minutes after your load stops?

**Reply by Zeng1998** [reply]
> Hi Zeng. Do you have a sense of what are the maximum number concurrent requests you have open at a time? Do you see the 100 number decay some minutes after your load stops?

I made about 100 concurrent requests at a time, and the Vercel function log show me that there are about 5 new connections of mongodb. But the mongodb cloud monitor show me that there are more than 100 connections at that time and it decays after finishing all requests.

**Reply by willshulman** [reply]
Hi Zeng. Thank you. I would expect Atlas to report 100 connections during the time you had 100 concurrent requests to the database, and for that number to decay as the load subsides, as you are seeing. @leerob may have more insight into what the information in the Vercel function logs signify.

**Reply by leerob** [reply]
It doesn't appear you're using the example mentioned and instead using express - I would recommend using Vercel Functions instead of a single Express instance, as mentioned here.

https://vercel.com/guides/using-express-with-vercel

**Reply by Zeng1998** [reply]
Thank you both! I will try to use Vercel function to do my project.
